### PR TITLE
Implement rb_enc_interned_str

### LINKED ISSUE #3703: rb_enc_interned_str() not yet implemented
But this is used e.g. in json: https://github.com/ruby/json/pull/675#issuecomment-2453071738
cc @byroot

### PR DESCRIPTION
**Pull request by Th3-M4jor**
closes #3703

### CODE REVIEW

**Comment by Th3-M4jor** [comment]
On an unrelated note that might just be an issue with my local setup but `bin/jt test cexts --no-gems` fails with
```
Error during command: test cexts --no-gems
/home/spartan364/truffleruby-ws/truffleruby/tool/jt.rb:676:in `bitbucket_url': Need a git remote in truffleruby with the internal repository URL (RuntimeError)
        from /home/spartan364/truffleruby-ws/truffleruby/tool/jt.rb:1755:in `gem_test_pack'
        from /home/spartan364/truffleruby-ws/truffleruby/tool/jt.rb:1512:in `block in run_single_cexts_test'
        from /home/spartan364/truffleruby-ws/truffleruby/tool/jt.rb:1583:in `time_test'
        from /home/spartan364/truffleruby-ws/truffleruby/tool/jt.rb:1469:in `run_single_cexts_test'
        from /home/spartan364/truffleruby-ws/truffleruby/tool/jt.rb:1461:in `block in test_cexts'
        from /home/spartan364/truffleruby-ws/truffleruby/tool/jt.rb:1460:in `each'
        from /home/spartan364/truffleruby-ws/truffleruby/tool/jt.rb:1460:in `test_cexts'
        from /home/spartan364/truffleruby-ws/truffleruby/tool/jt.rb:1189:in `test'
        from /home/spartan364/truffleruby-ws/truffleruby/tool/jt.rb:3275:in `main'
        from /home/spartan364/truffleruby-ws/truffleruby/tool/jt.rb:3284:in `<main>'
```

File: CHANGELOG.md [hunk 1]
```diff
@@ -17,6 +17,7 @@ Compatibility:
 * Set `RbConfig::CONFIG['archincludedir']` (#3396, @andrykonchin).
 * Support the index/length arguments for the string argument to `String#bytesplice` added in 3.3 (#3656, @rwstauner).
 * Implement `rb_str_strlen()` (#3697, @Th3-M4jor).
+* Implement `rb_enc_interned_str()` (#3706, @Th3-M4jor).
```
Comment (andrykonchin) [inline CHANGELOG.md:20]: Shouldn't it be #3703?

**Review by andrykonchin — APPROVED** [review]
Looks very good. Could you please squash commits into a single one?

**Comment by andrykonchin** [comment]
> On an unrelated note that might just be an issue with my local setup but bin/jt test cexts --no-gems fails with

Yeah, it seems a bug. Thank you for reporting!

File: CHANGELOG.md (hunk 1 above)
Comment (eregon) [reply CHANGELOG.md:20]: It should, we use issue numbers in the changelog, not PR numbers.

**Comment by Th3-M4jor** [comment]
I've squashed the commits and fixed the changelog entry to point to the issue instead of the PR

**Pull request merged**